aws/resource_aws_gamelift_build_test.go

```diff
@@ -0,0 +1,302 @@
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSGameliftBuildDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSGameliftBuildBasicConfig(buildName, bucketName, zipPath, roleName, policyName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGameliftBuildExists("aws_gamelift_build.test", &conf),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "name", buildName),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "operating_system", "WINDOWS_2012"),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.#", "1"),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.0.bucket", bucketName),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.0.key", "tf-acc-test-gl-build.zip"),
+					resource.TestMatchResourceAttr("aws_gamelift_build.test", "storage_location.0.role_arn", roleArnRe),
+				),
+			},
+			{
+				Config: testAccAWSGameliftBuildBasicUpdateConfig(uBuildName, bucketName, zipPath, roleName, policyName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGameliftBuildExists("aws_gamelift_build.test", &conf),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "name", uBuildName),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "operating_system", "WINDOWS_2012"),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.#", "1"),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.0.bucket", bucketName),
+					resource.TestCheckResourceAttr("aws_gamelift_build.test", "storage_location.0.key", "tf-acc-test-gl-build.zip"),
+					resource.TestMatchResourceAttr("aws_gamelift_build.test", "storage_location.0.role_arn", roleArnRe),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAWSGameliftBuildExists(n string, res *gamelift.Build) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No Gamelift Build ID is set")
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).gameliftconn
+
+		req := &gamelift.DescribeBuildInput{
+			BuildId: aws.String(rs.Primary.ID),
+		}
+		out, err := conn.DescribeBuild(req)
+		if err != nil {
+			return err
+		}
+
+		b := out.Build
+
+		if *b.BuildId != rs.Primary.ID {
+			return fmt.Errorf("Gamelift Build not found")
+		}
+
+		*res = *b
+
+		return nil
+	}
+}
+
+func testAccCheckAWSGameliftBuildDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).gameliftconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_gamelift_build" {
+			continue
+		}
+
+		req := gamelift.DescribeBuildInput{
+			BuildId: aws.String(rs.Primary.ID),
```

Comment:
Do we need to check `rs.Primary.ID` beforehand? e.g.

```go
if rs.Primary.ID == "" {
  continue
}
```

Reply:
I'd say hopefully not? I mean we don't do it in in any `CheckDestroy` function for any other acceptance test and I'm hoping core would've already swiped the resource away from state if it doesn't have ID, if not then we have much bigger problems IMO.